From ca1aef42a5f8ab33d1b89bde18348110cfbcb689 Mon Sep 17 00:00:00 2001
From: Luiz Capitulino <lcapitulino@redhat.com>
Date: Thu, 10 Jul 2014 14:15:25 -0400
Subject: [mm] hugetlb: update_and_free_page(), don't clear PG_reserved bit

Message-id: <1405001727-22802-7-git-send-email-lcapitulino@redhat.com>
Patchwork-id: 85661
O-Subject: [RHEL7.1 PATCH 6/8] hugetlb: update_and_free_page(): don't clear PG_reserved bit
Bugzilla: 996763
RH-Acked-by: Rik van Riel <riel@redhat.com>
RH-Acked-by: Andrea Arcangeli <aarcange@redhat.com>

Hugepages pages never get the PG_reserved bit set, so don't clear it.

However, note that if the bit gets mistakenly set free_pages_check() will
catch it.

Signed-off-by: Luiz Capitulino <lcapitulino@redhat.com>
Reviewed-by: Davidlohr Bueso <davidlohr@hp.com>
Acked-by: Kirill A. Shutemov <kirill.shutemov@linux.intel.com>
Reviewed-by: Zhang Yanfei <zhangyanfei@cn.fujitsu.com>
Cc: Andrea Arcangeli <aarcange@redhat.com>
Cc: David Rientjes <rientjes@google.com>
Cc: Marcelo Tosatti <mtosatti@redhat.com>
Cc: Naoya Horiguchi <n-horiguchi@ah.jp.nec.com>
Cc: Rik van Riel <riel@redhat.com>
Cc: Yasuaki Ishimatsu <isimatu.yasuaki@jp.fujitsu.com>
Cc: Yinghai Lu <yinghai@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
(cherry picked from commit a7407a27c2bba3711d272d72d2d63ea147a929df)
Signed-off-by: Luiz Capitulino <lcapitulino@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/mm/hugetlb.c b/mm/hugetlb.c
index e275cb6..2dbb223 100644
--- a/mm/hugetlb.c
+++ b/mm/hugetlb.c
@@ -565,8 +565,8 @@ static void update_and_free_page(struct hstate *h, struct page *page)
  for (i = 0; i < pages_per_huge_page(h); i++) {
   page[i].flags &= ~(1 << PG_locked | 1 << PG_error |
     1 << PG_referenced | 1 << PG_dirty |
-    1 << PG_active | 1 << PG_reserved |
-    1 << PG_private | 1 << PG_writeback);
+    1 << PG_active | 1 << PG_private |
+    1 << PG_writeback);
  }
  VM_BUG_ON(hugetlb_cgroup_from_page(page));
  set_compound_page_dtor(page, NULL);
-- 
1.7.1